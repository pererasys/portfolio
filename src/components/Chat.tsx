/**
 * @author Andrew Perera
 * Copyright (C) 2021
 */

import React, { useEffect, useState } from "react";

import {
  faArrowRight,
  faEllipsisH,
  faEnvelope,
  faPhone,
  faTimes,
  faCommentAlt,
} from "@fortawesome/free-solid-svg-icons";
import { BeatLoader } from "react-spinners";

import { enableBodyScroll, disableBodyScroll } from "body-scroll-lock";

import { Button, IconAction, IconButton } from "./common/Buttons";
import Dropdown, { DropdownList } from "./common/Dropdown";
import Overlay from "./common/Overlay";

import styles from "../styles/components/Chat.module.scss";

interface IMessage {
  message: string;
}

const Message = (props: IMessage) => (
  <li className={styles.message}>{props.message}</li>
);

interface IMessageList {
  loading: boolean;
}

const MessageList = ({ loading }: IMessageList) => {
  const [messages, setMessages] = useState([
    "Ask me anything! eg. Where did you grow up?",
  ]);

  const renderMessage = (msg: string, index: number) => (
    <Message message={msg} key={String(index)} />
  );

  return (
    <ul className={styles.messageList}>
      {loading ? (
        <BeatLoader loading={loading} color="#d3d3d3" size={12} />
      ) : (
        messages.map(renderMessage)
      )}
    </ul>
  );
};

const NewMessage = () => (
  <div className={styles.new}>
    <textarea
      id="message"
      name="message"
      className={styles.newMessage}
      placeholder="What do you want to know?"
    ></textarea>
    <div className={styles.actions}>
      <Button label="Send" icon={faArrowRight} size="sm" color="accent" />
      <Dropdown
        renderAnchor={(open) => (
          <IconAction
            icon={faEllipsisH}
            size="sm"
            color="accent"
            onClick={open}
          />
        )}
        y="top"
        x="right"
      >
        <DropdownList
          options={[
            { label: "Email", icon: faEnvelope },
            { label: "Phone", icon: faPhone },
          ]}
        />
      </Dropdown>
    </div>
  </div>
);

const Chat = () => {
  const [active, setActive] = useState(false);

  useEffect(() => {
    const elem = document.querySelector("body");
    if (active) {
      disableBodyScroll(elem);
    } else {
      enableBodyScroll(elem);
    }
  }, [active]);

  const toggleChat = () => {
    setActive(!active);
  };

  const chatClasses = [styles.chat];
  const actionClasses = [styles.action];

  if (active) chatClasses.push(styles.active);
  else actionClasses.push(styles.active);

  return (
    <React.Fragment>
      <div className={chatClasses.join(" ")}>
        <div className={styles.header}>
          <h4>Chat</h4>
          <IconAction
            icon={faTimes}
            color="accent"
            onClick={toggleChat}
            size="sm"
          />
        </div>
        <MessageList loading={true} />
        <NewMessage />
      </div>
      <div className={actionClasses.join(" ")}>
        <IconButton icon={faCommentAlt} color="accent" onClick={toggleChat} />
      </div>
      {active && <Overlay onClick={toggleChat} />}
    </React.Fragment>
  );
};

export default Chat;
