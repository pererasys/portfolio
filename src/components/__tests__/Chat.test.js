import { create } from "react-test-renderer";

import Chat from "../Chat";

describe("<Chat />", () => {
  const setup = () => {
    const component = create(<Chat />);

    return {
      component,
    };
  };

  it("should render without crashing", () => {
    const { component } = setup();

    expect(component.toJSON()).toMatchSnapshot();
  });
});
