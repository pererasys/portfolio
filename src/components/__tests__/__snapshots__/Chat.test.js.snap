// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chat /> should render without crashing 1`] = `
Array [
  <div
    className=""
  >
    <div>
      <h4>
        Chat
      </h4>
      <a
        onClick={[Function]}
      >
        <svg
          aria-hidden="true"
          className="svg-inline--fa fa-times fa-w-11 "
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          style={Object {}}
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
            style={Object {}}
          />
        </svg>
      </a>
    </div>
    <ul>
      <span
        className="css-0"
      >
        <span
          className="css-1euoelx"
        />
        <span
          className="css-1uguell"
        />
        <span
          className="css-1euoelx"
        />
      </span>
    </ul>
    <div>
      <textarea
        id="message"
        name="message"
        placeholder="What do you want to know?"
      />
      <div>
        <button>
          Send
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-arrow-right fa-w-14 "
            data-icon="arrow-right"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M190.5 66.9l22.2-22.2c9.4-9.4 24.6-9.4 33.9 0L441 239c9.4 9.4 9.4 24.6 0 33.9L246.6 467.3c-9.4 9.4-24.6 9.4-33.9 0l-22.2-22.2c-9.5-9.5-9.3-25 .4-34.3L311.4 296H24c-13.3 0-24-10.7-24-24v-32c0-13.3 10.7-24 24-24h287.4L190.9 101.2c-9.8-9.3-10-24.8-.4-34.3z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
        <div>
          <a
            onClick={[Function]}
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-ellipsis-h fa-w-16 "
              data-icon="ellipsis-h"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M328 256c0 39.8-32.2 72-72 72s-72-32.2-72-72 32.2-72 72-72 72 32.2 72 72zm104-72c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72zm-352 0c-39.8 0-72 32.2-72 72s32.2 72 72 72 72-32.2 72-72-32.2-72-72-72z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </a>
        </div>
      </div>
    </div>
  </div>,
  <div
    className=" "
  >
    <a
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-comment-alt fa-w-16 "
        data-icon="comment-alt"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M448 0H64C28.7 0 0 28.7 0 64v288c0 35.3 28.7 64 64 64h96v84c0 9.8 11.2 15.5 19.1 9.7L304 416h144c35.3 0 64-28.7 64-64V64c0-35.3-28.7-64-64-64z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
    </a>
  </div>,
]
`;
